Mesh split

arovinelli opened the pull request:
## Design
A new MeshModifier that split a monolithic mesh by adding nodes at the interface between different blocks.
In the case of multiple blocks present in the mesh, the user has two options:

- generate one single interface
- break the new interface by block ID (this is done setting the `split_interface=true` . In this case boundary id is automatically selected and the boundary name are assigned as bM#_bS#

ref #11775

lindsayad commented on framework/include/meshmodifiers/BreakMeshByBlock.h:
```diff
@@ -0,0 +1,53 @@
+public:
+  BreakMeshByBlock(const InputParameters & parameters);
+
+  virtual void modify();
```
This should also be marked `override`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/207857/) on b9f4682 wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/11797/site)

This comment will be updated on new commits.

arovinelli commented:
@lindsayad there is something going wild with this branch. it makes BISON fail to compile. it seems like my branch is corrupted.  I tried a

```
git fetch upstream
git reset upstream/devel
```

but it didn't work
any idea?

permcody commented:
Ignore the BISON failure. There's an incompatible update making it's way through CIVET.

arovinelli commented:
@permcody 
There some fixes required for distributes mesh... I was expecting this as I didn't coded at all the distributed mesh part.
On the other hand I'm a bit surprised about the failure of recover. Does the recover file use both the cpr and exodus. If so I'm, in trouble here as I'm adding node on the fly.

jiangwen84 commented:
@arovinelli I noticed that you used some my old codes that I took from an open source code (https://sourceforge.net/projects/cigen/) with a GPLv2 license. We should at least add comments about the origins of those functions. @permcody is there any other license issue? Thanks!

permcody commented:
@arovinelli For a new capability like this it's also fine to just not support the DistributedMesh capability. We can put in the normal error and disable those tests. I plan to have discussions with the rest of the team on how we want to handle this capability. I think there are several ways that this can/will break so I want to make sure we've thought through this carefully. Most of the team is out this week on vacation with the holiday falling in the middle of the week.

permcody commented:
@jiangwen84 - We definitely need to mark the cigen code appropriately. I'd be willing to add it as a contrib if that makes sense. I haven't really looked to see how you used it, or how much of it.

arovinelli commented:
@jiangwen84 tfor my there is no problem in adding a citation to the website. 
Let me know what is your standard way of adding this things into a source file.
If you want I can contact the author CCing @permcody  and asking for permission of reuse of some portion of the code (don't know how much is left really)

permcody commented:
Oh GPL... Actually you guys can't commit that to the MOOSE repository. It's incompatible with LGPL. The only way that this can go in is if we contact the developers and ask for permission to use the software under LGPL.

jiangwen84 commented:
If I remember correctly, there are four functions: buildNodeSupport(), buildInterfacialNodes(), duplicateNodes(), tearElements() in the BreakMeshByBlock class which are copied from cigen code. We should contact the developer and ask for his permission.

arovinelli commented:
@permcody 
here's the difference between the code @jiangwen84 and the current: [https://github.com/arovinelli/moose/compare/MeshSplit...arovinelli:MeshSplit2](url)
Yes we need to ask permission as the main concept behind the main loops are from his code.
If you are all ok with this, I can write him and ask for permission  and ccing @permcody 
such that the same reply goes to everyone.

arovinelli commented:
@permcody and @jiangwen84 as you should have seen from the emails, we received permission from Dr. Nguyen to freely utilize and distribute the code available [https://sourceforge.net/projects/cigen/](url)

permcody commented:
Yes, I received it. I'll make sure we have the appropriate documentation before we merge this. Thanks for contacting him.

friedmud commented on framework/doc/content/documentation/systems/MeshModifiers/BreakMeshByBlock.md:
```diff
@@ -0,0 +1,36 @@
+!listing test/tests/mesh_modifiers/break_mesh_by_blocks/2D/TestBreakMesh_2DJunction_splitTrue_Auto.i block=MeshModifiers
+
+The naming convention for the interface is `bMi_bSj`, where $i$ and $j$ represents the original blocks ids.
+This convention also assumes that the master interface $i$ is set on the block with lower id, while the slave interface $j$ is always on the block with higher id.
```
Is this (the block numbering dependency) a requirement of this object to work?

arovinelli replied:
Derek, if you are talking about `bMi_bSj` it's just the way the object names the new generate sidesets for split_interface=true. This is an output.

I made the choice to generate sidesets always on the MasterSide (lower block id) because you don't need complicated checks and each block can have as many neighbours as you want.

Right now you can't selectively choose blocks to split, the object will just make interface s between all the neighboring block (selective splitting is not trivial)


Does this answer your question?

arovinelli replied:
@friedmud I was thinking to your question while driving home, and I think this is the answer. No we don't need this convention`bMi_bSj`. If block names are guaranteed to be unique we can use block names. For instance if we have two blocks `steel` and `concrete` with ids `1` and `2`,respectively, we can name the interface something like `steel_concrete_interface`.
Is this what you want? Also do you like the suffix `_interface` or you want something more descriptive like `_splitinterface`.

friedmud commented on framework/doc/content/documentation/systems/MeshModifiers/BreakMeshByBlock.md:
```diff
@@ -0,0 +1,36 @@
+
+!listing test/tests/mesh_modifiers/break_mesh_by_blocks/2D/TestBreakMesh_2DJunction_splitTrue_Auto.i block=MeshModifiers
+
+The naming convention for the interface is `bMi_bSj`, where $i$ and $j$ represents the original blocks ids.
```
Can you provide an example so that this can be clearer?

arovinelli replied:
For instance if block 1 is  neighboring blocks 2 and 3 you will have two new interfaces:
bM1_bS2 and bM1_bS3 both sidesets will be added to block one being them the one with the lower id. 
I will add this to the documentaion

friedmud reviewed (changes requested):
Ok - overall - this seems to be too much code for what's going on here.  I think the reason is that you're trying to use an externally developed algorithm/code for something that is actually not that hard to do in MOOSE/libMesh.

If you look at @lindsayad 's recent PR where he's creating lower dimensional elements from a sideset... you can see more of what this code should look like: https://github.com/idaholab/moose/pull/11800/files#diff-0666ac4d8e83869853f528c53b556fbeR52

A lot of the things you're building here are things that we already have cached.

For instance, you should be using `MooseMesh::nodeToElemMap()` to find which elements are connected to the nodes.

If I were writing this algorithm it would go something like this:

```
for all_nodes
  connected_elems = node_to_elem_map[node->id()]
  block_multiplicity = findBlockMultiplicity(connected_elems)
  for i < block_multiplicity - 1
    new_node = new Node(node)
    mesh.add_node(new_node)
    fixup_elem = connecte_elems[i]
    which_node_to_fix = whichNodeAmI(elem, node)
    elem.set_node(which_node_to_fix) = new_node
```

That should really be all that's necessary.  Just walk through the nodes shared by multiple blocks, create duplicates and then fix the elements connected to it.

But maybe I'm simplifying the problem a bit too much... I haven't tried to code it myself...

arovinelli commented:
@friedmud You know MOOSE and libmesh  much more than. Your comments are great as I keep learning things. I'm not sure the algorithm you posted would work (need to check what happens at N-node junctions and make sure each added node is alway assigned to the same block) but I can try to improve the code following your suggestions. I'll get back to you in a few days.
In the meanwhile I have question: are neighbor informations stored in checkpoint files? If not how do we make restart work for this kind of MeshModifier?

friedmud commented:
You don't need to worry about restart - the "broken up" mesh will be
written out correctly - and will be re-read as a broken up mesh.  The
MeshModifiers are not run for recovery at all.

Derek

On Thu, Jul 5, 2018 at 3:59 PM Andrea Rovinelli <notifications@github.com>
wrote:

> @friedmud <https://github.com/friedmud> You know MOOSE and libmesh much
> more than. Your comments are great as I keep learning things. I'm not sure
> the algorithm you posted would work (need to check what happens at N-node
> junctions and make sure each added node is alway assigned to the same
> block) but I can try to improve the code following your suggestions. I'll
> get back to you in a few days.
> In the meanwhile I have question: are neighbor informations stored in
> checkpoint files? If not how do we make restart work for this kind of
> MeshModifier?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/11797#issuecomment-402865208>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA1JMe0NUE7oMRUWsk0_yKhQKjAUaYzzks5uDowugaJpZM4VBgj4>
> .
>

friedmud commented:
My algorithm handles the multiplicity of multiple blocks coming together...

On Thu, Jul 5, 2018 at 4:18 PM Derek Gaston <friedmud@gmail.com> wrote:

> You don't need to worry about restart - the "broken up" mesh will be
> written out correctly - and will be re-read as a broken up mesh.  The
> MeshModifiers are not run for recovery at all.
>
> Derek
>
> On Thu, Jul 5, 2018 at 3:59 PM Andrea Rovinelli <notifications@github.com>
> wrote:
>
>> @friedmud <https://github.com/friedmud> You know MOOSE and libmesh much
>> more than. Your comments are great as I keep learning things. I'm not sure
>> the algorithm you posted would work (need to check what happens at N-node
>> junctions and make sure each added node is alway assigned to the same
>> block) but I can try to improve the code following your suggestions. I'll
>> get back to you in a few days.
>> In the meanwhile I have question: are neighbor informations stored in
>> checkpoint files? If not how do we make restart work for this kind of
>> MeshModifier?
>>
>> —
>> You are receiving this because you were mentioned.
>>
>>
>> Reply to this email directly, view it on GitHub
>> <https://github.com/idaholab/moose/pull/11797#issuecomment-402865208>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AA1JMe0NUE7oMRUWsk0_yKhQKjAUaYzzks5uDowugaJpZM4VBgj4>
>> .
>>
>

arovinelli commented:
@friedmud I tried to use all the function available in MOOSE and libmesh and used the cycle you suggested now the code is much shorter.

friedmud commented on framework/src/meshmodifiers/BreakMeshByBlock.C:
```diff
@@ -0,0 +1,170 @@
+  for (auto node_it = node_to_elem_map.begin(); node_it != node_to_elem_map.end(); ++node_it)
+  {
+    const dof_id_type current_node_id = node_it->first;
+    const Node current_node = _mesh_ptr->getMesh().node(current_node_id);
```
Please save off a `libMesh::Mesh &` (what comes back from `getMesh()`) so you're not calling that all over the place.

friedmud commented on framework/src/meshmodifiers/BreakMeshByBlock.C:
```diff
@@ -0,0 +1,170 @@
+
+    // find node multiplicity
+    const std::set<SubdomainID> connected_blocks = _mesh_ptr->getNodeBlockIds(current_node);
+    unsigned int node_multiplicity = connected_blocks.size();
```
Nice!

friedmud commented on framework/src/meshmodifiers/BreakMeshByBlock.C:
```diff
@@ -0,0 +1,170 @@
+                  current_node.id()) // if current node == node on element
+              {
+                current_elem->set_node(local_elem_node_it) = new_node;
+                break; // we re cycling over nodes, therefore we can fix only one node a ta  time
```
Couple of typos

friedmud commented on framework/src/meshmodifiers/BreakMeshByBlock.C:
```diff
@@ -0,0 +1,170 @@
+          // check if we need to fix a node in the current element
+          // if current_elem belong to the reference subdomain
+          if (current_elem->subdomain_id() == reference_subdomain_id)
+            reference_elems.insert(current_elem->id()); // store reference eleemnt
```
`eleemnt`

friedmud reviewed (dismissed):
This looks much better!  Well done!

Just a couple of small changes...

permcody commented on test/src/mesh_modifiers/BreakMeshByBlockManual_3Blocks.C:
```diff
@@ -0,0 +1,196 @@
+#include "BreakMeshByBlockManual_3Blocks.h"
+#include "MooseMesh.h"
+
+registerMooseObject("MooseApp", BreakMeshByBlockManual_3Blocks);
```
Make sure you register this with the MooseTestApp, not MooseApp. You will then have to add a flag to your test spec `allow_test_objects = true`

arovinelli commented on framework/src/meshmodifiers/BreakMeshByBlock.C:
```diff
@@ -0,0 +1,170 @@
+
+  // initialize the node to elemen map
+  const std::map<dof_id_type, std::vector<dof_id_type>> & node_to_elem_map =
+      _mesh_ptr->nodeToElemMap();
```
@friedmud Should we explicitly call `MooseMesh::update()` at the end of the main loop to reinit the `_node_to_element_map` or it is done automatically at the end of each mesh modifier?

permcody replied:
I'd have to double check, but I'm going to say don't call it. We try to minimize calls to reinit and prepare because they can get expensive so MOOSE generally does all of this when it's needed. If you run into a problem, let us know.

friedmud replied:
I honestly don't know - can you go put a print in the nodeToElemMap() function (where it gets rebuilt) to make sure that it _does_ get rebuilt after your modifier?

I think it does - but always good to check

arovinelli replied:
@friedmud 
I added I print statement at the end of the `BreakMeshByBlock::modify` and one iniside `MooseMesh::update`

the result is  the following
```
MooseMesh::update() reset _node_to_elem_map
END OF BreakMeshByBlock
MooseMesh::update() reset _node_to_elem_map
```

So it gets resetted at some point

friedmud replied:
👍

permcody commented on framework/src/meshmodifiers/BreakMeshByBlockBase.C:
```diff
@@ -0,0 +1,139 @@
+      (_pars.isParamSetByUser("interface_id") || _pars.isParamSetByUser("interface_name")))
+  {
+    mooseError("if split_interface == true, interface_name and/or interface_id"
+               " cannot be specified by the user");
```
By the user, or at all?

permcody commented on framework/src/meshmodifiers/BreakMeshByBlockBase.C:
```diff
@@ -0,0 +1,139 @@
+
+  // check if the boundary name already exist
+  bool checkBoundaryAlreadyExist = false;
+  for (auto b : _bName_bID_set)
```
👍

permcody commented on framework/src/meshmodifiers/BreakMeshByBlockBase.C:
```diff
@@ -0,0 +1,139 @@
+                                             BoundaryID & boundaryID,
+                                             BoundaryInfo & boundary_info)
+{
+  // TODO this method should be multiprocessor and multithread safe
```
This will work with ReplicatedMesh without any special code but needs work for DistributedMesh (you don't have to necessarily make this work with DistributedMesh, that can often be a chore).

arovinelli commented on framework/include/meshmodifiers/BreakMeshByBlockBase.h:
```diff
@@ -0,0 +1,72 @@
+  /// the name of the new interface
+  std::string _interface_name;
+  /// the ID of the new interface
+  BoundaryID _interface_id;
```
deleted the ID

arovinelli commented on framework/src/meshmodifiers/BreakMeshByBlock.C:
```diff
@@ -0,0 +1,170 @@
+    const Node current_node = mesh.node(current_node_id);
+
+    // find node multiplicity
+    const std::set<SubdomainID> connected_blocks = _mesh_ptr->getNodeBlockIds(current_node);
```
done

arovinelli commented on framework/src/meshmodifiers/BreakMeshByBlock.C:
```diff
@@ -0,0 +1,170 @@
+    {
+
+      // retrieve connected elements from the map
+      std::vector<dof_id_type> connected_elems = node_it->second;
```
done

arovinelli commented on framework/src/meshmodifiers/BreakMeshByBlock.C:
```diff
@@ -0,0 +1,170 @@
+      // the original node remains on the reference_subdoamin
+      for (auto subdomain_it = std::next(connected_blocks.begin());
+           subdomain_it != connected_blocks.end();
+           ++subdomain_it)
```
now I understand what a range loop is!!
Done

arovinelli commented on framework/src/meshmodifiers/BreakMeshByBlock.C:
```diff
@@ -0,0 +1,170 @@
+        // fix all the elements belonging to the connected_blocks[subdomain_it]
+        for (unsigned int elem_it = 0; elem_it < connected_elems.size(); elem_it++)
+        {
+          Elem * current_elem = mesh.elem(connected_elems[elem_it]);
```
done

arovinelli replied:
done

arovinelli commented on framework/src/meshmodifiers/BreakMeshByBlock.C:
```diff
@@ -0,0 +1,170 @@
+        mesh.add_node(new_node);
+
+        // fix all the elements belonging to the connected_blocks[subdomain_it]
+        for (unsigned int elem_it = 0; elem_it < connected_elems.size(); elem_it++)
```
completely changed: now it loops over multiplicity

arovinelli replied:
again... done

arovinelli commented on framework/src/meshmodifiers/BreakMeshByBlock.C:
```diff
@@ -0,0 +1,170 @@
+  for (auto node_it = node_to_elem_map.begin(); node_it != node_to_elem_map.end(); ++node_it)
+  {
+    const dof_id_type current_node_id = node_it->first;
+    const Node current_node = mesh.node(current_node_id);
```
checked, even though I'm not completely if I should raise an error if the pointer is null

arovinelli commented on framework/src/meshmodifiers/BreakMeshByBlock.C:
```diff
@@ -0,0 +1,259 @@
+}
+
+void
+BreakMeshByBlock::buildInterfacialNodes()
```
this doesn't exist anymore

permcody commented:
OK - I looked at this a little bit and think I understand what's happening here. In general, we don't support documenting test objects through MooseDocs. What we'd like to do is create good design documentation for new interfaces in the framework/modules and create test objects that exercise those interfaces. The design shouldn't be about the test object though, it should be able the interfaces or objects that you added. 

There's actually a good reason for what we are doing. We want to be able to build the documentation completely offline so that somebody can go back in history and build the documentation at a past revision. Making sure that there's good integrity for the actual portions of the framework/modules you are using is required to make this work correctly. 

The solution here is to simply remove your documentation to your test object. Yes, please keep good code comments, but don't worry about the MD files and references to those files through your "design" parameters.

Commit:
add new MeshModifier to split a mesh by block: `BreakMeshByBlock` ref #11775

arovinelli commented:
@permcody thanks for the reply. I update the test files as suggested.
Also I addressed all the comments you did last night and updated the code accordingly.

Thanks

mlesueur commented:
Hi @arovinelli ,
Congrats on the pull request!
I saw that your tests are done with a FileMesh. Have you tested the mesh modifier on a generated mesh?
I have ran the input file written under and I don't seem to be able to create the interface.

The input file is super simple:
[Mesh]
  type = GeneratedMesh
  dim = 2
  ny = 2
[]

[MeshModifiers]
  [subdomain]
    type = SubdomainBoundingBox
    bottom_left = '0 0.5 0'
    block_id = 1
    top_right = '1 1 0'
  []
  [interface]
    type = BreakMeshByBlock
    depends_on = 'subdomain'
  []
[]

friedmud commented:
@mlesueur what do you want it to do with a GeneratedMesh?  A GeneratedMesh only has one block!

friedmud commented:
Never mind - I see what you’re trying to do this after another MeshModifier.  My guess is that the results of that meshmodifier are possibly not getting cached in MooseMwsh...
